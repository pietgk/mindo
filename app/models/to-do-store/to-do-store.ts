import KSUID from "ksuid"
import { Instance, SnapshotOut, types } from "mobx-state-tree"
import { ToDoId, ToDoModel, ToDoState, withEnvironment } from ".."

export const KSUid = () => KSUID.randomSync().string as Id;
export type Id = string;

export const ToDoStoreModel = types
  .model("ToDoStore")
  .props({
    todos: types.optional(types.map(ToDoModel), {}),
  })
  .extend(withEnvironment)
  // .views((self) => ({}))
  .actions((self) => ({
    addToDo(task: string) {
      const id = KSUid()
      const todo = ToDoModel.create({ id, task, state: ToDoState.ACTIVE }) 
      self.todos.put(todo)
      return todo
    },
    completed(id: ToDoId) {
      self.todos.get(id).setState(ToDoState.COMPLETED)
    },
    delete(id: ToDoId) {
      self.todos.get(id).setState(ToDoState.DELETED)
    }
  }))

type ToDoStoreType = Instance<typeof ToDoStoreModel>
export interface ToDoStore extends ToDoStoreType { }
type ToDoStoreSnapshotType = SnapshotOut<typeof ToDoStoreModel>
export interface ToDoStoreSnapshot extends ToDoStoreSnapshotType { }
export const createToDoStoreDefaultModel = () => types.optional(ToDoStoreModel, {})
