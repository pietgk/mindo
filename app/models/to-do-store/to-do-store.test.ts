import { getSnapshot } from "mobx-state-tree"
import { ToDoStoreModel } from "./to-do-store"

describe("ToDoStore", () => {
  it("can be created", () => {
    const instance = ToDoStoreModel.create({})
    expect(getSnapshot(instance)).toMatchInlineSnapshot(`
      Object {
        "todos": Object {},
      }
    `)
  })
})

describe("ToDoStore User", () => {
  it("can add a new Task to the TO-DO list", () => {
    const instance = ToDoStoreModel.create({})
    const todo = instance.addToDo("buy milk")
    expect(getSnapshot(instance.todos.get(todo.id))).toMatchSnapshot({
      id: expect.any(String),
    })
  })
  it("can get all Tasks from the TO-DO list for visualisation", () => {
    const instance = ToDoStoreModel.create({})
    const tasks = ["buy milk", "feed the cat"]
    tasks.forEach((task) => instance.addToDo(task))
    instance.todos.forEach(task => {
      expect(task).toMatchSnapshot({
        id: expect.any(String)
      })
    })
  })
  it("can mark a Task in the TO-DO list as complete", () => {
    const instance = ToDoStoreModel.create({})
    const todo = instance.addToDo("buy milk")
    instance.completed(todo.id)
    expect(getSnapshot(instance.todos.get(todo.id))).toMatchSnapshot({
      id: expect.any(String),
    })
  })
  it("can delete a Task from the TO-DO list", () => {
    const instance = ToDoStoreModel.create({})
    const todo = instance.addToDo("buy milk")
    instance.delete(todo.id)
    expect(getSnapshot(instance.todos.get(todo.id))).toMatchSnapshot({
      id: expect.any(String),
    })
  })
})
