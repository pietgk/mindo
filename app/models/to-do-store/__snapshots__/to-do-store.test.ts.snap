// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoStore User can add a new Task to the TO-DO list 1`] = `
Object {
  "id": Any<String>,
  "state": "🟢",
  "task": "buy milk",
}
`;

exports[`ToDoStore User can delete a Task from the TO-DO list 1`] = `
Object {
  "id": Any<String>,
  "state": "❌",
  "task": "buy milk",
}
`;

exports[`ToDoStore User can get all Tasks from the TO-DO list for visualisation 1`] = `
Object {
  "id": Any<String>,
  "state": "🟢",
  "task": "buy milk",
}
`;

exports[`ToDoStore User can get all Tasks from the TO-DO list for visualisation 2`] = `
Object {
  "id": Any<String>,
  "state": "🟢",
  "task": "feed the cat",
}
`;

exports[`ToDoStore User can mark a Task in the TO-DO list as complete 1`] = `
Object {
  "id": Any<String>,
  "state": "✅",
  "task": "buy milk",
}
`;
